tests/data_replicator.rs

```diff
@@ -183,3 +183,105 @@ pub fn crdt_retransmit() {
         t.join().unwrap();
     }
 }
+
+#[test]
+fn check_external_liveness_table() {
+    logger::setup();
+    let c1_c4_exit = Arc::new(AtomicBool::new(false));
+    let c2_c3_exit = Arc::new(AtomicBool::new(false));
+
+    trace!("c1:");
+    let (c1, dr1, _) = test_node(c1_c4_exit.clone());
+    trace!("c2:");
+    let (c2, dr2, _) = test_node(c2_c3_exit.clone());
+    trace!("c3:");
+    let (c3, dr3, _) = test_node(c2_c3_exit.clone());
+    trace!("c4:");
+    let (c4, dr4, _) = test_node(c1_c4_exit.clone());
+
+    let c1_data = c1.read().unwrap().my_data().clone();
+    c1.write().unwrap().set_leader(c1_data.id);
+
+    let c2_id = c2.read().unwrap().me;
+    let c3_id = c3.read().unwrap().me;
+    let c4_id = c4.read().unwrap().me;
+
+    // Insert the remote data about c4
+    let c2_index_for_c4 = 10;
+    c2.write().unwrap().remote.insert(c4_id, c2_index_for_c4);
+    let c3_index_for_c4 = 20;
+    c3.write().unwrap().remote.insert(c4_id, c3_index_for_c4);
+
+    // Set up the initial network topology
+    c2.write().unwrap().insert(&c1_data);
+    c3.write().unwrap().insert(&c1_data);
+
+    c2.write().unwrap().set_leader(c1_data.id);
+    c3.write().unwrap().set_leader(c1_data.id);
+
+    // Wait to converge
+    trace!("waiting to converge:");
+    let mut done = false;
+    for _ in 0..30 {
+        done = c1.read().unwrap().table.len() == 3 && c2.read().unwrap().table.len() == 3
```

Comment:
Old version of `cargo fmt` - this will fail CI. If you `rustup update stable`, it'll automatically update rustfmt along with the compiler.